OpenRA.Game/Network/OrderManager.cs

```diff
@@ -170,59 +226,180 @@ void CheckSync(byte[] packet)
+			PerfHistory.Increment("net_slowdown", slowDownRequired < 0 ? slowDownRequired : 0);
+			PerfHistory.Increment("net_simlag", 100.0 * (double)SimLag / (double)SimLagLimit);
+			PerfHistory.Increment("net_speed", 100.0 * (double)World.Timestep / (double)ActualTimestep);
 		}
 
-		public IEnumerable<Session.Client> GetClientsNotReadyForNextFrame
+		IEnumerable<Session.Client> GetClientsNotReadyForNextFrame
 		{
 			get
 			{
 				return GameStarted
-					? frameData.ClientsNotReadyForFrame(NetFrameNumber)
+					? frameData.ClientsNotReadyForFrame()
 						.Select(a => LobbyInfo.ClientWithIndex(a))
 					: NoClients;
 			}
 		}
 
-		public void Tick()
+		void SendOrders()
 		{
-			if (!IsReadyForNextFrame)
-				throw new InvalidOperationException();
+			if (!GameStarted)
+				return;
 
-			if (GameSaveLastFrame < NetFrameNumber + FramesAhead)
-				Connection.Send(NetFrameNumber + FramesAhead, localOrders.Select(o => o.Serialize()).ToList());
+			if (GameSaveLastFrame < NextOrderFrame)
+			{
+				Connection.Send(NextOrderFrame, localOrders.Select(o => o.Serialize()).ToList());
+				localOrders.Clear();
+			}
 
-			localOrders.Clear();
+			NextOrderFrame++;
+		}
 
-			foreach (var order in frameData.OrdersForFrame(World, NetFrameNumber))
+		/*
+		 * Only available if TickImmediate() is called first and we are ready to dispatch received orders locally.
+		 * Process all incoming orders for this frame, handle sync hashes and step our net frame.
+		 */
+		void ProcessOrders()
+		{
+			var orders = frameData.OrdersForFrame(World).ToList();
+			foreach (var order in orders)
 				UnitOrders.ProcessOrder(this, World, order.Client, order.Order);
 
-			if (NetFrameNumber + FramesAhead >= GameSaveLastSyncFrame)
+			if (NetFrameNumber % SyncFrameScale == 0)
 			{
-				var defeatState = 0UL;
-				for (var i = 0; i < World.Players.Length; i++)
-					if (World.Players[i].WinState == WinState.Lost)
-						defeatState |= 1UL << i;
+				if (NetFrameNumber >= GameSaveLastSyncFrame)
+				{
+					var defeatState = 0UL;
+					for (var i = 0; i < World.Players.Length; i++)
+						if (World.Players[i].WinState == WinState.Lost)
+							defeatState |= 1UL << i;
+
+					Connection.SendSync(NetFrameNumber, OrderIO.SerializeSync(World.SyncHash(), defeatState));
+				}
+				else
+					Connection.SendSync(NetFrameNumber, OrderIO.SerializeSync(0, 0));
 
-				Connection.SendSync(NetFrameNumber, OrderIO.SerializeSync(World.SyncHash(), defeatState));
+				if (generateSyncReport)
+					using (new PerfSample("sync_report"))
+						syncReport.UpdateSyncReport(orders);
 			}
-			else
-				Connection.SendSync(NetFrameNumber, OrderIO.SerializeSync(0, 0));
 
-			if (generateSyncReport)
-				using (new PerfSample("sync_report"))
-					syncReport.UpdateSyncReport();
+			if (frameData.TryPeekTimestep(out var timestep))
+				ActualTimestep = timestep;
+
+			frameData.AdvanceFrame();
 
 			++NetFrameNumber;
 		}
 
+		public void TickPreGame()
```

Comment:
There are two things that don't sit well with me here:
1. The name describes the thing that calls this rather than what it actually does. We should rename this to something more descriptive or, better, fold it into a special case within `TryTick`.
2. I think that the logic here is also going to be needed for fixing the `ShouldUseCatchUp` issue i commented in Game above.